gpii/node_modules/gpii-oauth2/gpii-oauth2-datastore/src/DbDataStoreUtils.js

```diff
@@ -0,0 +1,408 @@
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = fluid || require("infusion"),
+    gpii = fluid.registerNamespace("gpii"),
+    $ = fluid.registerNamespace("jQuery"),
+    uuid = require("node-uuid");
+
+require("../../gpii-oauth2-utilities/src/OAuth2Utilities.js");
+
+fluid.registerNamespace("gpii.oauth2.dbDataStore");
+
+// All doc types used for saving different records into CouchDB/PouchDB
+gpii.oauth2.dbDataStore.docTypes = {
+    user: "user",
+    gpiiToken: "gpiiToken",
+    client: "client",
+    authDecision: "authDecision",
+    authCode: "authCode",
+    clientCredentialsToken: "clientCredentialsToken"
+};
+
+gpii.oauth2.dbDataStore.findRecord = function (dataSource, termMap, valueNotEmpty, dataProcessFunc) {
+    dataProcessFunc = dataProcessFunc || gpii.oauth2.dbDataStore.CleanUpDoc;
+    var promiseTogo = fluid.promise();
+
+    // Verify required field values are provided instead of undefined
+    var emptyFields = gpii.oauth2.dbDataStore.verifyEmptyFields(termMap, valueNotEmpty);
+
+    if (emptyFields.length > 0) {
+        var error = gpii.oauth2.composeError(gpii.oauth2.errors.missingInput, {fieldName: emptyFields});
+        promiseTogo.reject(error);
+    } else {
+        var promise = dataSource.get(termMap);
+        promise.then(function (data) {
+            // $.isEmptyObject() is to work around the issue when fetching data
+            // using pouch/couch DB views and records are not found, instead of
+            // returning a 404 status code, it returns this object:
+            // { total_rows: 1, offset: 0, rows: [] }
+            // Note the "rows" value is an empty array.
+            // This behavior prevents "kettle.dataSource.CouchDB" -> "notFoundIsEmpty"
+            // option from returning "undefined". Instead, an empty object {}
+            // is returned. This work around is to make sure "undefined" is returned
+            // when an empty object is received.
+            var result = $.isEmptyObject(data) ? undefined : dataProcessFunc(data);
+            if (result !== undefined && result.isError) {
+                promiseTogo.reject(result);
+            } else {
+                promiseTogo.resolve(result);
+            }
+        }, function (error) {
+            promiseTogo.reject(error);
+        });
+    }
+
+    return promiseTogo;
+};
+
+gpii.oauth2.dbDataStore.verifyEmptyFields = function (termMap, valueNotEmpty) {
+    var emptyFields = "",
+        count = 0;
+
+    valueNotEmpty = fluid.makeArray(valueNotEmpty);
+    fluid.each(valueNotEmpty, function (fieldName) {
+        if (termMap[fieldName] === undefined) {
+            emptyFields = count === 0 ? fieldName : emptyFields + " & " + fieldName;
+            count++;
+        }
+    });
+    return emptyFields;
+};
+
+// Remove CouchDB/PouchDB internal fields: _id, _rev and type. Also save "_id" filed value into "id" field.
+gpii.oauth2.dbDataStore.CleanUpDoc = function (data) {
```

Comment:
"_id" is mapped to "id" is to maintain the API backward compatibility as the data store API is expected to output the record identifier in "id" field instead of a couchdb/pouchdb specific name of "_id".

This explanation is also added into the code comment.